### PR TITLE
[4.3] Rewriting com_tags router

### LINKED ISSUE #13598: [4.0]TAGS: router doesn't reuse exsting routes
### Steps to reproduce the issue
create some menu items pointing to either an article or a category blog
attach some tags to the articles
then create a tagged menu item
check the results: 
### Expected result
the routes on each result should reuse the routes from the menu items we have created 

### Actual result
the routes generated look like id-category/id-alias instead of using existing routes

### System information (as much as possible)
J3.6.5

### LINKED ISSUE #35382: [4.0] Paging Problem Tags
### Steps to reproduce the issue
If you have 100's of Articles Tagged with the same Tag, like "Year 2021" and you want to see them, with 10 Articles per Page.

The First Page displays Correctly.

The Second Page returns 404 not found.

The First Page is linked like this:
/index.php/en/component/tags/tag/year-2021

The second page like this:
/index.php/en/component/tags/year-2021?Itemid=237&start=5

With "/tag" missing. => result 404.

### Expected result

"/tag" is included.

### Actual result

"/tag" is missing.

### System information (as much as possible)

J 4.0.2 upgrade from 3.10.1

### Additional comments

This works correctly on J 3.10.1

### LINKED ISSUE #37444: J4.1.2 raises error 404 when clicking on tag starting with a numeric character in the output of menutype "List All Tags"
### Steps to reproduce the issue
Create a tag named "TestParentTag", no parent
Create a tag named "alpha2020", parent: "TestParentTag"
Create a tag named "2020", parent: "TestParentTag"

Create a menu item of type "List All Tags", parent tag: "TestParentTag"
Go to frontend, to the new menu item created, in the tag list, click on "alpha2020" Output: as expected
Go to frontend, to the new menu item created, in the tag list, click on "2020" Output: 404 error page

### Expected result
List of articles with the tag 

### Actual result
Error 404 page

### System information (as much as possible)
See attached file

### Additional comments
Issue arose after updating from J4.1.0 to J4.1.2
[systeminfo-2022-04-01T11_09_48+02_00.txt](https://issues.joomla.org/uploads/1/d29b62b098985bd678ee4b5f8ebab47d.txt)

### LINKED ISSUE #38996: ?layout=list prefix in the URL for compact list of tagged items
Here is the structure:

1. menu item. domain.com/tags/
Tags. Compact list of tagged items
Снимок экрана 2022-10-19 в 17 05 51

2. Settings. The menu item settings include various tags
 
Снимок экрана 2022-10-19 в 17 26 49

I use tags to navigate though tagged articles. There is no sense to make menu item for each tag since tagged items solves the task.
However, I noticed a little problem with router in breadcrumbs:

Home = domain.com/
Tags = domain.com/tags?layout=list
Tagged item = domain.com/tags/plugin

I suggest considering of removing **?layout=list** from the URL if a menu item exist. A change in Joomla router.
Joomla 4.2.3

Thank you.

### PR DESCRIPTION
**Pull request by Hackwar**
Pull Request for Issues #39105, #38996, #35382, #13598, #37444, #37492.

### Summary of Changes
This PR rewrites the router of com_tags. Above you can see a whole list of issues which this should fix. Since these are very different issues, please don't close the issues right away without first testing this if it really fixes the issue.

This router improves a lot of things, but it also still is not perfect. While this better matches existing menu items, removes unnecessary URL parameters, allows for multiple tags per URL and fixes bugs in pagination and a few other areas, this still does not properly support language and multilanguage setups. It also fails to properly support nested tags.

The routes of the new router should be pretty much identical to the existing ones, except for where the current URLs aren't actually working. Right now the router can generate and parse URLs with several tags in one URL, where the IDs of the tags are handed over as an array in the non-SEF URL and a SEFed URL will have the aliases of those tags concatenated as segments of the URL.

### Stuff not solved by this PR
The multilanguage issues are serious and I would love to solve them, but those fixes would build upon these changes here and these changes then again are already complex enough to test, so I'd like to first merge this here and then fix the language issues in another PR, especially since that PR would require a lot of changes to the models of the component.

The nested tags issue in the end means a new feature. Right now you can set a tag to be a child of another tag. However, URLs to a tag only contain the alias, not the complete path of the tag. This would be problematic when 2 child tags of different parents have the same alias. There is also no visual representation of this structure. My proposal would be to introduce an option to switch between the ability to select more than one tag per URL by using /tags/[tag1]/[tag2]/... or selecting a single, nested tag by using /tags/[tag1-level1]/[tag1-level2]/...

Another new feature could be to add a plus sign to tags in order to further filter a list of items down by adding yet another tag to that filter. Right now several tags in a menu item don't mean that only items with all tags associated to these items are shown, but all items which have at least one of those tags are displayed. 

#### Pagination issue (fixes #35382 & #39105)
1. Install testing sampledata
2. Unpublish all tag related menu items (search for "tag" in the "all menu items" backend view)
3. Create a tag which you then assign to at least 21 articles
4. In the frontend, lookup one of the articles and click on the new tag
5. In the single tag view for that tag, click on the second page.
6. Without this change, this should result in an error. With this change, this should be fine.

#### Additional URL parameter (fixes #38996)
Please go to #38996 and follow the instructions there to rebuild the issue. After applying this change, this should be fixed.

#### Use the right menu item for tags (fixes #13598)
1. Install testing sampledata
2. Create two tags and create a menu item for one of the two with a single tag view
3. Assign the new tags to an article
4. Due to the "All Tags" menu item, after applying the change, the tag with the menu item should point to that menu item and the other should point to the all-tags menu item

#### Numeric tags don't work (fixes #37444)
1. Install testing sampledata
2. Create a tag with a numeric title
3. Assign it to an article
4. Try to click on that tag in the article in the frontend. Without this change, you don't get a valid URL, with this change, the right page is displayed.

#### Additional URL parameter v2 (#37492)
I can't reproduce this with this change applied. Can someone check if this is valid?

#### Improvements with these changes
1. You can now have URLs with multiple tags. Such a non-SEFed URL would look like this: `index.php?option=com_tags&view=tag&id[0]=1:alias-1&id[1]=2:alias-2...`
2. SEFed URLs with multiple tags would look like this: `/[menu-item]/[tag-1]/[tag-2]/...`

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

I don't know right now if we need documentation changes...

### CODE REVIEW

**Comment by richard67** [comment]
Setting the release blocker label as inherited from issue #39105 .

File: components/com_tags/src/Helper/RouteHelper.php [hunk 1]
```diff
@@ -97,38 +98,34 @@ public static function getTagRoute($id)
...
+     * @param   string|string[]   $id        The ID of the tag in the format TAG_ID:TAG_ALIAS
+     * @param   string            $language  The language of the tag
      *
      * @return  string  URL link to pass to the router
      *
      * @since   4.2.0
      * @throws  Exception
      */
-    public static function getComponentTagRoute(string $id, string $language = '*'): string
+    public static function getComponentTagRoute($id, string $language = '*'): string
```
Comment (bembelimen) [inline components/com_tags/src/Helper/RouteHelper.php]: This change will break B/C

**Review by bembelimen — DISMISSED** [review]

**Comment by jamfx** [comment]
I have tested this item :white_check_mark: successfully on 8f23f4f480e5fa069e8155fcbe671ed6dae070ce<br /><br />Followed the detailed instructions. Works as described. No errors found.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39114">issues.joomla.org/tracker/joomla-cms/39114</a>.</sub>

File: components/com_tags/src/Helper/RouteHelper.php (hunk 1 above)
Comment (Hackwar) [reply components/com_tags/src/Helper/RouteHelper.php]: Why does this break B/C? It is less restrictive than before...

Comment (Hackwar) [reply components/com_tags/src/Helper/RouteHelper.php]: Reverted

**Comment by chmst** [comment]
Test on a fresh installation (4.3 dev) with testing sample data and additional own data, not multilingual on localhost (sef urls on but no rewrite rule).

Could not replicate before applying the patch (successful per se)
Pagination issue (fixes #35382 & #39105)
Numeric tags don't work (fixes #37444)

Could not replicate: Additional URL parameter v2 (#37492)

Successful 
Additional URL parameter (fixes #38996)
Use the right menu item for tags (fixes #13598)

Looks good for me but recommend testing with multilingual sites. No documentation change needed. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39114">issues.joomla.org/tracker/joomla-cms/39114</a>.</sub>

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on a3cd27e294c224c701108fc99beb225d0edc5e93<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39114">issues.joomla.org/tracker/joomla-cms/39114</a>.</sub>

**Comment by richard67** [comment]
I've restored @jamfx 's test result since all changes after it were not really functional.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39114">issues.joomla.org/tracker/joomla-cms/39114</a>.</sub>

**Comment by obuisard** [comment]
Thank you Hannes @Hackwar for this much needed PR!

**Comment by ch2856** [comment]
J4.3.0
The new router adds ?Itemid=104 to all general tags (no menu)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39114">issues.joomla.org/tracker/joomla-cms/39114</a>.</sub>

**Comment by joeforjoomla** [comment]
Confirmed. The new router has broken routing adding ?Itemid=xxx to all links despite the menu item created.

**Comment by NicolasDerumigny** [comment]
I am unsure whether I should open an issue for it or just a comment, but when:
- SEF Url is enable
- a menu entry exists for "list all tags (with root as parent)", setting for example the alias `tags` for `index.php?option=com_tags&view=tags&parent_id=1`

Then tag URL are not rewritten from `mysite.com/component/tags/tag/name_of_tag` to `mysite.com/tags/name_of_tag`, though the latter is also valid URL and behave as expected.

Is this the correct behavior?

**Comment by Fedik** [comment]
We have a PR that for fixing broken behavior of Tag router  #40474. Please test, and report your result there. Thanks.

**Pull request merged**